src/objects.c

```diff
@@ -3186,3 +3938,51 @@ CK_RV p11prov_merge_pub_attrs_into_priv(P11PROV_OBJ *pub_key,
     P11PROV_raise(priv_key->ctx, ret, "Failed attr copy");
     return CKR_GENERAL_ERROR;
 }
+
+/* creates an empty (no public point) Public EC Key, (OpenSSL paramgen
+ * function), that will later be filled in with a public EC key obtained
+ * by a peer */
+P11PROV_OBJ *mock_pub_ec_key(P11PROV_CTX *ctx, CK_ATTRIBUTE_TYPE type,
+                             CK_ATTRIBUTE *ec_params)
+{
+    P11PROV_OBJ *key;
+    CK_RV ret;
+
+    key = p11prov_obj_new(ctx, CK_UNAVAILABLE_INFORMATION, CK_INVALID_HANDLE,
+                          CK_UNAVAILABLE_INFORMATION);
+    if (!key) {
+        return NULL;
+    }
+
+    key->class = CKO_PUBLIC_KEY;
+    key->data.key.type = type;
+
+    /* at this stage we have EC_PARAMS  and the preprocessing
+     * function will add CKA_P11PROV_CURVE_NID and
+     * CKA_P11PROV_CURVE_NAME */
+    key->attrs = OPENSSL_zalloc(KEY_EC_PARAMS * sizeof(CK_ATTRIBUTE));
+    if (key->attrs == NULL) {
+        P11PROV_raise(key->ctx, CKR_HOST_MEMORY,
+                      "Failed to generate mock ec key");
+        p11prov_obj_free(key);
+        return NULL;
+    }
+
+    ret = p11prov_copy_attr(&key->attrs[0], ec_params);
+    if (ret != CKR_OK) {
+        P11PROV_raise(key->ctx, CKR_HOST_MEMORY, "Failed to generate mock key");
+        p11prov_obj_free(key);
+        return NULL;
+    }
+    key->numattrs++;
+
+    /* verify params are ok */
+    ret = pre_process_ec_key_data(key);
+    if (ret != CKR_OK) {
+        P11PROV_raise(key->ctx, CKR_HOST_MEMORY, "Failed to generate mock key");
```

Comment:
I would prefer to have different error messages  that would be more informative (or even different return values) so finding the origin of the error would be easier when there will be different versions, patches and code lines wont match.

Reply:
ok